Fix shell fs mounting for block device and add support of h3ulcb to shell fs sample

xakep-amatop opened the pull request.

xakep-amatop committed:
fixup! WA: fs: littlefs: support block devices in shell

Fix mounting of littlefs blk device. lfs_mount stucks in case when
the read/prog buffer less than SDMMC block size, see how function
'card_read_blocks' works.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>
Acked-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>

xakep-amatop committed:
samples: shell: fs: add rcar_h3ulcb_ca57 support to shell/fs sample

Add config and overlay file for rcar_h3ulcb_ca57 board support to
subsys/shell/fs sample.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>
Acked-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>

xakep-amatop commented on subsys/fs/shell.c:
```diff
@@ -535,15 +535,6 @@ static int cmd_mount_fat(const struct shell *shell, size_t argc, char **argv)
 
 static int cmd_mount_littlefs(const struct shell *shell, size_t argc, char **argv)
 {
-#ifdef CONFIG_FS_LITTLEFS_BLK_DEV
```
these changes aren't needed at all

xakep-amatop commented on subsys/fs/shell.c:
```diff
@@ -36,7 +36,7 @@ static struct fs_mount_t fatfs_mnt = {
 /* TODO: Implement dynamic storage dev selection */
 #ifdef CONFIG_FS_LITTLEFS_BLK_DEV
-FS_LITTLEFS_DECLARE_DEFAULT_CONFIG(lfs_data);
+static struct fs_littlefs lfs_data = {0};
```
mount of littlefs for blk device fails in case when we use `FS_LITTLEFS_DECLARE_DEFAULT_CONFIG` initializer
make the same initializator as for little fs sample

oleksiimoisieiev replied:
As I can see FS_LITTLEFS_DECLARE_DEFAULT_CONFIG creates not only fs_littlefs structure, but read_buffer, prog_buffer and lookahead_buffer and cachesize = 0.  FS_LITTLEFS_DECLARE_CUSTOM_CONFIG comment says that application can define and initialize struct fs_littlefs directly, but application is responsible to ensure the configuration. If I understood correctly you had to set CONFIG_FS_LITTLEFS_FC_HEAP_SIZE in config because you are using cachesize = 0 in lfs_data. I'm not sure if it's correct. As for me - I suggest to make 2 things: 1) Make lfs_data initialization for each driver (move it under ifdef) and use FS_LITTLEFS_DECLARE_CUSTOM_CONFIG to set correct values. If it's impossible and lfs_data should be 0 for both drivers - write a comment with the desctiption why it's so.

xakep-amatop replied:
I just copied and pasted parts of code from LittleFS samples for block devices and haven't delved deep into it, it works correctly. The problem was with the `FS_LITTLEFS_DECLARE_DEFAULT_CONFIG` --- the LittleFS mount call getting stuck. I'll investigate your concerns.

xakep-amatop replied:
changed to ` FS_LITTLEFS_DECLARE_CUSTOM_CONFIG` and add description to commit message why it helps

oleksiimoisieiev replied:
+1

dsemenets commented:
Looks good for me

oleksiimoisieiev commented on subsys/fs/shell.c:
```diff
@@ -36,12 +36,21 @@ static struct fs_mount_t fatfs_mnt = {
+
+#if defined(CONFIG_DISK_DRIVER_SDMMC)
+#define DISK_NAME CONFIG_SDMMC_VOLUME_NAME
+#elif IS_ENABLED(CONFIG_DISK_DRIVER_MMC)
```
```
#if defined(.....
....
#elif IS_ENABLED(...
```
I think IS_ENABLED macro should be used for both cases. Just to be consistent.

xakep-amatop replied:
it is a common approach:
```
samples/subsys/fs/littlefs/src/main.c:313:#if defined(CONFIG_DISK_DRIVER_SDMMC)
samples/subsys/fs/littlefs/src/main.c-314-#define DISK_NAME CONFIG_SDMMC_VOLUME_NAME
samples/subsys/fs/littlefs/src/main.c-315-#elif IS_ENABLED(CONFIG_DISK_DRIVER_MMC)
samples/subsys/fs/littlefs/src/main.c-316-#define DISK_NAME CONFIG_MMC_VOLUME_NAME
samples/subsys/fs/littlefs/src/main.c-317-#else
samples/subsys/fs/littlefs/src/main.c-318-#error "No disk device defined, is your board supported?"
--
tests/drivers/disk/disk_access/src/main.c:17:#if defined(CONFIG_DISK_DRIVER_SDMMC)
tests/drivers/disk/disk_access/src/main.c-18-#define DISK_NAME CONFIG_SDMMC_VOLUME_NAME
tests/drivers/disk/disk_access/src/main.c-19-#elif IS_ENABLED(CONFIG_DISK_DRIVER_MMC)
tests/drivers/disk/disk_access/src/main.c-20-#define DISK_NAME CONFIG_MMC_VOLUME_NAME
tests/drivers/disk/disk_access/src/main.c-21-#elif IS_ENABLED(CONFIG_DISK_DRIVER_RAM)
tests/drivers/disk/disk_access/src/main.c-22-#define DISK_NAME CONFIG_DISK_RAM_VOLUME_NAME
--
tests/drivers/disk/disk_performance/src/main.c:15:#if defined(CONFIG_DISK_DRIVER_SDMMC)
tests/drivers/disk/disk_performance/src/main.c-16-#define DISK_NAME CONFIG_SDMMC_VOLUME_NAME
tests/drivers/disk/disk_performance/src/main.c-17-#elif IS_ENABLED(CONFIG_DISK_DRIVER_MMC)
tests/drivers/disk/disk_performance/src/main.c-18-#define DISK_NAME CONFIG_MMC_VOLUME_NAME
tests/drivers/disk/disk_performance/src/main.c-19-#elif IS_ENABLED(CONFIG_DISK_DRIVER_RAM)
tests/drivers/disk/disk_performance/src/main.c-20-#define DISK_NAME CONFIG_DISK_RAM_VOLUME_NAME
```

oleksiimoisieiev replied:
..... Or was written this way once and the copied to all places. 
Anyway it works botch ways, but looks weird and produces a question is there some hidden meaning in using different checks in #if and #elif...
I'm okay with current approach. It's up to you whether change this or not.

xakep-amatop replied:
changed to `#elif defined`

oleksiimoisieiev replied:
+1

lorc reviewed (commented):
I do understand that "fixup!" is not complete patch. But nevertheless, it is good to have some description as about what was wrong and how it was fixed.

This will greatly help person who will squash your changes into the final version of the patch, because they will need to update the commit message accordingly to changes in the fixup.

xakep-amatop committed:
fs: littlefs: add alignment arg to FS_LITTLEFS_DECLARE_CUSTOM_CONFIG macro

Add alignment parameter to FS_LITTLEFS_DECLARE_CUSTOM_CONFIG macro, it
speeds up read/write operation for SDMMC devices in case when we align
buffers on CONFIG_SDHC_BUFFER_ALIGNMENT, because we can avoid extra copy
of data from card bffer to read/prog buffer. See, how 'card_read_blocks'
function works.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>
Acked-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>

xakep-amatop commented:
> I do understand that "fixup!" is not complete patch. But nevertheless, it is good to have some description as about what was wrong and how it was fixed.
> 
> This will greatly help person who will squash your changes into the final version of the patch, because they will need to update the commit message accordingly to changes in the fixup.

added detailed description

oleksiimoisieiev commented on include/zephyr/fs/littlefs.h:
```diff
@@ -62,15 +62,17 @@ struct fs_littlefs {
  *
  * @param name the name for the structure.  The defined object has
  * file scope.
+ * @param alignment needed alignment for reads/prog buffer for specific device
```
s/reads/read/g

oleksiimoisieiev commented on samples/subsys/shell/fs/boards/rcar_h3ulcb_ca57.conf:
```diff
@@ -0,0 +1,12 @@
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+CONFIG_FS_LITTLEFS_FC_HEAP_SIZE=8192
```
I believe you don't need fc_heap_size param anymore. As I've mentioned here https://github.com/xen-troops/zephyr/pull/67#discussion_r1266345135 "you had to set CONFIG_FS_LITTLEFS_FC_HEAP_SIZE in config because you are using cachesize = 0 in lfs_data". Now you have set cachesize so i think this param can be removed.

xakep-amatop replied:
Actually, the sample crashes without this param

oleksiimoisieiev replied:
Ok.

oleksiimoisieiev commented:
Some minor comments, but:
LGTM
`Acked-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>`

firscity commented on subsys/fs/shell.c:
```diff
@@ -10,6 +10,7 @@
 #include <zephyr/shell/shell.h>
```
Commit message: 
nit: second and third sentences may be splitted to simpler :) 
"th3e" -> "the"

firscity commented on samples/subsys/shell/fs/boards/rcar_h3ulcb_ca57.overlay:
```diff
@@ -0,0 +1,8 @@
+/* SPDX-License-Identifier: Apache-2.0 */
```
EPAM copyright?

xakep-amatop replied:
as far as I see, it isn't a common approach, but Zephyr project has dozens overlay files without copyright

firscity reviewed (approved):
With comments fixed:
`Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>`

xakep-amatop commented:
fixed comments and added acks

firscity reviewed (approved):
`Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>`
